README.md
# GrowthViz

GrowthViz was developed in partnership between the Health FFRDC and CDC, with
feedback from leading health researchers, to support post-processing and data
visualization of growthcleanr output.

The objective of this tool is to allow users to conduct post-processing and data
visualization of growthcleanr output.
[growthcleanr](https://github.com/carriedaymont/growthcleanr) is an automated
method for cleaning longitudinal anthropometric data from EHRs. GrowthViz
provides an environment that includes graphical user interfaces as well as
interactive software development to explore data.

## Contents

- [Git Repository Information](#git-repository-information)
- [GrowthViz Purpose](#growthviz-purpose)
- [Background](#background)
- [Simple Install](#simple-install)
- [Sample data and first run testing](#sample-data-and-first-run-testing)
- [Docker Install](#docker-install)

## Git Repository Information

The latest code for this project should run `GrowthViz-pediatrics.ipynb` or
`GrowthViz-adults.ipynb`, depending on the user's patient population.

The notebook requires Python 3, Jupyter Notebook, Pandas, Matplotlib and
Seaborn. Some widgets also require the Qgrid extension enabled in Jupyter. The
`.csv` files in the repository are the source data required to run the notebook.
Custom data should replace these files in the same format. For more details see
[the simple install instructions below.](#simple-install)

## GrowthViz Purpose

The objective of this tool is to allow users to conduct post-processing and data
visualization of growthcleanr output.
[growthcleanr](https://github.com/carriedaymont/growthcleanr) is an automated
method for cleaning longitudinal anthropometric growth data from EHRs. It is
available as open source software. GrowthViz is to be used **after** a data set
has been run through growthcleanr.

### Background

As stated in
[Automated identification of implausible values in growth data from pediatric electronic health records](https://academic.oup.com/jamia/article/24/6/1080/3767271):

> In pediatrics, evaluation of growth is fundamental, and many pediatric
> research studies include some aspect of growth as an outcome or other
> variable. The clinical growth measurements obtained in day-to-day care are
> susceptible to error beyond the imprecision inherent in any anthropometric
> measurement. Some errors result from minor problems with measurement
> technique. While these errors can be important in certain analyses, they are
> often small and generally impossible to detect after measurements are
> recorded. Larger measurement technique errors can result in values that are
> biologically implausible and can cause problems for many analyses.

GrowthViz uses data sets that were evaluated with growthcleanr. The tool expects
the output to be in a CSV format that is described later on in the notebook.

GrowthViz is a [Juypter Notebook](https://jupyter.org/). It provides an
environment that includes graphical user interfaces as well as interactive
software development to explore data. To achieve this, GrowthViz references
different software languages and packages:

- [Python programming language](https://www.python.org/) is used to import,
  transform, visualize and analyze the output of growthcleanr. Some of the code
  for the tool is directly included in this notebook. Other functions have been
  placed in an external file to minimize the amount of code that users see in
  order to let them focus on the actual data.

- Data analysis is performed using [NumPy](https://numpy.org/) and
  [Pandas](https://pandas.pydata.org/). The output of growthcleanr will be
  loaded into a
  [pandas DataFrame](https://pandas.pydata.org/pandas-docs/stable/reference/api/pandas.DataFrame.html).
  GrowthViz provides functions for transforming DataFrames to support
  calculation of some values, such as BMI, as well as supporting visualizations.
  It is expected that users will create views into or copies of the DataFrames
  built initially by this tool. Adding columns to the DataFrames created by this
  tool is unlikely to cause problems. Removing columns is likely to break some
  of the tool's functionality.

- Visualization in the tool is provided by [Matplotlib](https://matplotlib.org/)
  and [Seaborn](http://seaborn.pydata.org/). Users may generate their own charts
  with these utilities.

## Simple Install

Anaconda is an all-in-one package installer for setting up dependencies needed
to run and view GrowthViz.

1. Install Anaconda

- Follow install instructions
  [found here for installation.](https://docs.anaconda.com/anaconda/install/)
- Opt for the Python 3.7 version
- The
  [windows install instructions](https://docs.anaconda.com/anaconda/install/windows/)
  are step-by-step and will get everything set up properly for the project.

2. Download the [GrowthViz project](https://github.com/mitre/GrowthViz) as a zip
   file using the "Clone or download" button on GitHub.

3. Unzip the GrowthViz zip file to have access to all of the source files for
   the Jupyter notebook.

4. Run the Anaconda Navigator that was installed during Step 1 (go to Start >
   Anaconda Navigator). This may take a while to load.

5. Before Launching the Jupyter Notebook application (shown on the home page),
   download one additional dependency "Qgrid". To do this:

- Click 'Environments' on the left.

- Type 'Qgrid' in the `Search Packages` text box in the top center of the
  screen. If it shows up with a green checkbox, proceed to Step 6.

- If it does not appear:

  - Change the 'Installed' drop down in the top center of the application to
    'Not Installed' and type in 'Qgrid' in the search bar on the right.

    - If Qgrid still does not show up click 'Update Index...' button next to the
      search bar. This may take several minutes. Once it is done search for
      Qgrid again.

  - Check the box to the left of Qgrid in the list and click the green 'Apply'
    button in the lower right corner.

  - Confirm the installation dialog. Installation may again take several
    minutes.

  - Once installation is successful, click on the 'Home' in the upper left
    navigation panel and proceed to Step 6.

6. Click ‘Launch’ under the ‘Jupyter Notebook’ icon. This will open the Jupyter
   Notebook interface in your default browser.

7. Within the browser, navigate to the `GrowthViz-master` folder you downloaded
   and unzipped in Step 2 (likely found in your Downloads/ folder). Click on
   `GrowthViz.ipynb` to run the Python notebook.

8. **[Optional step for testing the notebook]** Once the notebook is open, click
   the 'Run' button to step through the various blocks (cells) of the document,
   OR click the 'Cell' dropdown in the menu bar and select 'Run all' to test the
   entire notebook all at once.

If not using Anaconda, specific versions of packages can be found in
`requirements.txt`.

## Sample data and first run testing

By default when you reach Step 6 of the [Simple Install](#simple-install)
instructions above the notebook will use sample data loaded from the `.csv`
files located in the GrowthViz project. This is the same synthetic sample data
that is packaged with
[growthcleanr](https://github.com/carriedaymont/growthcleanr), cleaned and then
separated into pediatric and adult sets for GrowthViz.

To ensure that all of the necessary example files are present, run the
`check_setup.py` script.

## Docker Install

Docker allows for the ability to download GrowthViz and its dependencies in an
environment. To use this method,
[download and install Docker Desktop](https://www.docker.com/products/docker-desktop).
It also requires a GitHub account and a GitHub (personal access
token](https://docs.github.com/en/authentication/keeping-your-account-and-data-secure/creating-a-personal-access-token).

1. Log in to GitHub (create an account if required).

2. In GitHub, create a
   [fine-grained personal access token](https://docs.github.com/en/authentication/keeping-your-account-and-data-secure/creating-a-personal-access-token#creating-a-fine-grained-personal-access-token).
   Note that these have an expiration date with a configurable term; you may
   have to regenerate this token.

3. Copy the token value and store it somewhere safe on your local machine.

4. On the command line, log in to GitHub's container registry:

```bash
docker login ghcr.io --username [your-GitHub-username]
```

5. At the `Password:` prompt, paste in your token value (created in step 2).

6. Download and run GrowthViz-Docker with the following command. This may take
   several minutes to download components, especially the first time.

```bash
docker run -it -p 8888:8888 -v [data-path]/growthviz-data:/usr/src/app/growthviz-data ghcr.io/mitre/growthviz/mitre/growthviz
```

- Replace the `[data-path]` with a directory path you choose on your local
  computer. For instance, choose: `~/Documents` which means that a folder named
  `/growthviz-data` will be created in the Documents folder. To input data into
  GrowthViz, add CSV files in this `/growthviz-data` folder.

- Note also that when mapping a folder on Windows, you may be prompted to
  confirm that you indeed want to "Share" the folder. This is a standard Windows
  security practice, and it is okay to confirm and proceed.

7. View GrowthViz

- After running the above command, several lines of text will appear. Choose the
  third URL in this text and navigate to it in a web browser.

- The URL should be in the format: `http://X.X.X.X:8888/?token=XXX...`

- Within the browser, click on the file `GrowthViz.ipynb`. This will open a new
  window with the GrowthViz Jupyter Notebook.

8. Run GrowthViz

- You can choose to either click the `Run` button to step through the various
  blocks (cells) of the document, OR click the 'Cell' dropdown in the menu bar
  and select 'Run all' to test the entire notebook all at once. However, this
  will run with the default sample data. Step 4 will explain how to use your own
  data.

9. Input Your Own Dataset CSVs

- To input your own data, drop a file `[name-of-your-file.csv]` into the
  `/growthviz-data` folder you created in step 1.

- Then, navigate to Cells 7 and 28 and replace:

```python
cleaned_obs = pd.read_csv("sample-data-cleaned.csv")
```

...with:

```python
cleaned_obs = pd.read_csv("growthviz-data/[name-of-your-file.csv]")
```

- Where `[name-of-your-file.csv]` is the input CSV file you placed in your
  `/growthviz-data` folder.

#### Output boxes

When you run all cells (see Step 8 above) `Out[#]:` boxes will appear in the
notebook below the `In[#]:` code cells. These outputs are the result of the
functioning code blocks on the data. The "Out" blocks will often be interactive
charts and graphs used to explore the growthcleanr data. Descriptions of each
`Out[#]:` block can be found in the text sections above the `In[#]:` blocks.

## Notice

Copyright &copy; 2020-2022 The MITRE Corporation.

Approved for Public Release; Distribution Unlimited. Case Number 19-2008


NEWS.md
# GrowthViz 3.0.0-infants-beta - 2023-03-14

## Added

- Support for infant subjects (#56, #61, #66)
- Automated container image builds for main/latest and tags (#59)
- Added dependabot checks for GitHub workflow action updates (#68)

## Changed

- Completed function documentation throughout (#65)
- Updated sample data to include excluded age ranges (#60)
- Bug fixes (#62, #63, #64)
- Added Python versions 3.10, 3.11 to CI workflow for testing (#58)

# GrowthViz 2.0.0 - 2021-06-30

## Added

- Support for reviewing cleaned adult observations based on updated growthcleanr
  2.0 (#4, #15, #19, #32, #35)
- Test suite verifying baseline operations (#9, #26, #40)
- NEWS.md file (#46)

## Changed

- Improvements to charts (#10, #11, #24)
- Updated synthetic data to include adults (#37)
- Code reorganized for modularity, PEP8, library and data now in separate dirs
  (#6, #7, #14, #23)
- Updated Dockerfile (#31)


processdata.py
import pandas as pd
import numpy as np
import matplotlib.pyplot as plt
import matplotlib as mpl
from IPython.display import FileLinks


def setup_individual_obs_df(obs_df):
    """
    Standardizes adults and pediatrics files for clean processing in GrowthViz notebooks
    """
    df = obs_df.copy()
    df.rename(columns={"clean_res": "clean_value"}, inplace=True)
    df["age"] = df["agedays"] / 365.25
    df.drop(columns=["agedays"], inplace=True)
    df["clean_cat"] = df["clean_value"].astype("category")
    df["include"] = df.clean_value.eq("Include")
    col_list = [
        "id",
        "subjid",
        "param",
        "measurement",
        "age",
        "sex",
        "clean_value",
        "clean_cat",
        "include",
    ]
    return df[col_list]


def setup_percentiles_adults(percentiles):
    """
    Processes adults percentiles from CDC
    """
    # expand decade rows into one row per year
    pct = percentiles[
        percentiles["Age (All race and Hispanic-origin groups)"] != "20 and over"
    ].copy()
    pct.loc[pct["Age_low"] == 20, "Age_low"] = 18
    range_col = pct.apply(lambda row: row.Age_high - row.Age_low + 1, axis=1)
    pct = pct.assign(range=range_col.values)
    dta = pd.DataFrame(
        (np.repeat(pct.values, pct["range"], axis=0)), columns=pct.columns
    )
    dta["count"] = dta.groupby(["Sex", "Measure", "Age_low", "Age_high"]).cumcount()
    dta["age"] = dta["Age_low"] + dta["count"]
    # add standard deviation and other values
    dta["sqrt"] = np.sqrt(pd.to_numeric(dta["Number of examined persons"]))
    dta["sd"] = dta["Standard error of the mean"] * dta["sqrt"]
    dta["Sex"] = dta.Sex.replace("Male", 0).replace("Female", 1)
    dta.rename(columns={"Measure": "param"}, inplace=True)
    dta.drop(
        columns=[
            "Age (All race and Hispanic-origin groups)",
            "Age_low",
            "sqrt",
            "Standard error of the mean",
            "Age_high",
            "range",
            "count",
            "Number of examined persons",
        ],
        inplace=True,
    )
    # smooth percentiles between X9-(X+1)1 (i.e., 29-31)
    dta["decade"] = np.where(dta["age"] == (round(dta["age"].astype(float), -1)), 1, 0)
    mcol_list = [
        "Mean",
        "sd",
        "P5",
        "P10",
        "P15",
        "P25",
        "P50",
        "P75",
        "P85",
        "P90",
        "P95",
    ]
    for col in mcol_list:
        dta[col] = np.where(
            (dta["decade"] == 1) & (dta["age"] < 110),
            (dta[col] + dta[col].shift(1)) / 2,
            dta[col],
        )
    dta.drop(columns={"decade"}, inplace=True)
    col_list = ["param", "Sex", "age"] + mcol_list
    dta = dta.reindex(columns=col_list)
    return dta


def setup_percentiles_pediatrics(percentiles_file):
    """
    Processes pediatrics percentiles from CDC
    """
    percentiles = pd.read_csv(
        percentiles_file,
        dtype={
            "Agemos": float,
            "P5": float,
            "P50": float,
            "P95": float,
            "L": float,
            "M": float,
            "S": float,
            "Sex": int,
        },
    )
    percentiles["age"] = percentiles["Agemos"] / 12
    # Values by CDC (1=male; 2=female) differ from growthcleanr
    # which uses a numeric value of 0 (male) or 1 (female).
    # This aligns things to the growthcleanr values
    percentiles["Sex"] = percentiles["Sex"] - 1
    return percentiles


def keep_age_range(df, mode):
    """
    Restricts patient data to acceptable age range for notebooks
    """
    obs_grp = df
    # create age buckets for chart
    def label_excl_grp(row):
        if mode == "adults":
            if row["age"] < 18:
                return "  Below 18 (Exclude)"
            if (row["age"] >= 18) & (row["age"] < 30):
                return " 18 to < 30"
            if (row["age"] >= 30) & (row["age"] < 40):
                return " 30 to < 40"
            if (row["age"] >= 40) & (row["age"] < 50):
                return " 40 to < 50"
            if (row["age"] >= 50) & (row["age"] < 60):
                return " 50 to < 60"
            if (row["age"] >= 60) & (row["age"] <= 65):
                return " 60 to 65"
            if (row["age"] > 65) & (row["age"] <= 80):
                return " > 65 to 80 (Not Recommended)"
            if row["age"] > 80:
                return "Above 80 (Exclude)"
        elif mode == "pediatrics":
            if row["age"] < 2:
                return "  Below 2 (Exclude)"
            if (row["age"] >= 2) & (row["age"] < 5):
                return " 02 to < 05"
            if (row["age"] >= 5) & (row["age"] < 8):
                return " 05 to < 08"
            if (row["age"] >= 8) & (row["age"] < 11):
                return " 08 to < 11"
            if (row["age"] >= 11) & (row["age"] < 14):
                return " 11 to < 14"
            if (row["age"] >= 14) & (row["age"] < 17):
                return " 14 to < 17"
            if (row["age"] >= 17) & (row["age"] <= 20):
                return " 17 to 20"
            if (row["age"] > 20) & (row["age"] <= 25):
                return " > 20 to 25 (Not Recommended)"
            if row["age"] > 25:
                return "Above 25 (Exclude)"

    label_excl_col = obs_grp.apply(lambda row: label_excl_grp(row), axis=1)
    obs_grp = obs_grp.assign(cat=label_excl_col.values)
    obs_grp = (
        obs_grp.groupby("cat")["subjid"]
        .count()
        .reset_index()
        .sort_values("cat", ascending=True)
    )
    # assign bar colors
    cat_list = obs_grp["cat"].values.tolist()
    color_list = []
    patterns = []
    for n in cat_list:
        if ("Below" in n) | ("Above" in n):
            color_list = color_list + ["C3"]
            patterns = patterns + ["x"]
        if ("to" in n) & ("Not" not in n):
            color_list = color_list + ["C0"]
            patterns = patterns + [""]
        if "Not" in n:
            color_list = color_list + ["orange"]
            patterns = patterns + ["/"]
    # create chart
    fig, ax1 = plt.subplots()
    obs_grp_plot = plt.bar(
        obs_grp["cat"],
        obs_grp["subjid"],
        color=color_list,
    )
    for bar, pattern in zip(obs_grp_plot, patterns):
        bar.set_hatch(pattern)
    plt.xticks(rotation=45, ha="right")
    ax1.get_yaxis().set_major_formatter(
        mpl.ticker.FuncFormatter(lambda x, p: format(int(x), ","))
    )
    # return specified age range
    if mode == "adults":
        return df[df["age"].between(18, 80, inclusive=True)]
    elif mode == "pediatrics":
        return df[df["age"].between(2, 25, inclusive=True)]


def setup_merged_df(obs_df):
    """
    Merges together weight and height data for calculating BMI
    """
    obs_df = obs_df.assign(height=obs_df["measurement"], weight=obs_df["measurement"])
    obs_df.loc[obs_df.param == "WEIGHTKG", "height"] = np.NaN
    obs_df.loc[obs_df.param == "HEIGHTCM", "weight"] = np.NaN
    heights = obs_df[obs_df.param == "HEIGHTCM"]
    weights = obs_df[obs_df.param == "WEIGHTKG"]
    merged = heights.merge(weights, on=["subjid", "age", "sex"], how="outer")
    only_needed_columns = merged.drop(
        columns=[
            "param_x",
            "measurement_x",
            "clean_value_x",
            "weight_x",
            "id_y",
            "param_y",
            "measurement_y",
            "clean_value_y",
            "height_y",
        ]
    )
    clean_column_names = only_needed_columns.rename(
        columns={
            "clean_cat_x": "height_cat",
            "include_x": "include_height",
            "height_x": "height",
            "clean_cat_y": "weight_cat",
            "include_y": "include_weight",
            "weight_y": "weight",
            "reason_y": "reason",
            "id_x": "id",
        }
    )
    clean_column_names["bmi"] = clean_column_names["weight"] / (
        (clean_column_names["height"] / 100) ** 2
    )
    clean_column_names["rounded_age"] = np.around(clean_column_names.age)
    clean_column_names["include_both"] = (
        clean_column_names["include_height"] & clean_column_names["include_weight"]
    )
    return clean_column_names


def exclusion_information(obs):
    """
    Provides a count and percentage of growthcleanr categories by measurement type (param).

    Parameters:
    obs: a DataFrame, in the format output by setup_individual_obs_df

    Returns:
    A DataFrame with the counts and percentages
    """
    exc = (
        obs.groupby(["param", "clean_cat"])
        .agg({"id": "count"})
        .reset_index()
        .pivot(index="clean_cat", columns="param", values="id")
    )
    exc["height percent"] = exc["HEIGHTCM"] / exc["HEIGHTCM"].sum() * 100
    exc["weight percent"] = exc["WEIGHTKG"] / exc["WEIGHTKG"].sum() * 100
    exc = exc.fillna(0)
    exc["total"] = exc["HEIGHTCM"] + exc["WEIGHTKG"]
    exc = exc[["HEIGHTCM", "height percent", "WEIGHTKG", "weight percent", "total"]]
    exc = exc.sort_values("total", ascending=False)
    return exc.style.format(
        {
            "HEIGHTCM": "{:.0f}".format,
            "height percent": "{:.2f}%",
            "WEIGHTKG": "{:.0f}".format,
            "weight percent": "{:.2f}%",
        }
    )


def label_incl(row):
    """
    Categorizes BMI calculations as Include, Implausible, or unable to calculate (Only Wt or Ht)
    """
    if row["include_both"] == True:
        return "Include"
    elif (row["weight_cat"] == "Implausible") | (row["height_cat"] == "Implausible"):
        return "Implausible"
    else:
        return "Only Wt or Ht"


def setup_bmi_adults(merged_df, obs):
    """
    Appends BMI data onto adults weight and height observations
    """
    data = merged_df[
        [
            "id",
            "subjid",
            "sex",
            "age",
            "rounded_age",
            "bmi",
            "weight_cat",
            "height_cat",
            "include_both",
        ]
    ]
    incl_col = data.apply(lambda row: label_incl(row), axis=1)
    data = data.assign(clean_cat=incl_col.values)
    data["param"] = "BMI"
    data["clean_value"] = data["clean_cat"]
    data.rename(columns={"bmi": "measurement"}, inplace=True)
    return pd.concat([obs, data])


def data_frame_names(da_locals):
    """
    Returns a list of dataframe names
    """
    frames = []
    for key, value in da_locals.items():
        if isinstance(value, pd.DataFrame):
            if key.startswith("_") is False:
                frames.append(key)
    return frames


def export_to_csv(da_locals, selection_widget, out):
    """
    Saves out csv file of dataframe
    """
    df_name = selection_widget.value
    da_locals[df_name].to_csv(
        "growthviz-data/output/{}.csv".format(df_name), index=False
    )
    out.clear_output()
    out.append_display_data(FileLinks("growthviz-data/output"))


def clean_swapped_values(merged_df):
    """
    This function will look in a DataFrame for rows where the height_cat and weight_cat are set to
    "Swapped-Measurements". It will then swap the height and weight values for those rows.
    It will also create two new columns: postprocess_height_cat and postprocess_weight_cat.
    The values for these columns is copied from the original categories except in the case where
    swaps are fixed when it is set to "Include-Fixed-Swap".

    Parameters:
    merged_df: (DataFrame) with subjid, height, weight, include_height and include_weight columns

    Returns:
    The cleaned DataFrame
    """
    merged_df["postprocess_height_cat"] = merged_df["height_cat"]
    merged_df["postprocess_height_cat"] = merged_df[
        "postprocess_height_cat"
    ].cat.add_categories(["Include-Fixed-Swap"])
    merged_df["postprocess_weight_cat"] = merged_df["weight_cat"]
    merged_df["postprocess_weight_cat"] = merged_df[
        "postprocess_weight_cat"
    ].cat.add_categories(["Include-Fixed-Swap"])
    merged_df.loc[
        merged_df["height_cat"] == "Swapped-Measurements", ["height", "weight"]
    ] = merged_df.loc[
        merged_df["height_cat"] == "Swapped-Measurements", ["weight", "height"]
    ].values
    merged_df.loc[
        merged_df["height_cat"] == "Swapped-Measurements", "postprocess_height_cat"
    ] = "Include-Fixed-Swap"
    merged_df.loc[
        merged_df["weight_cat"] == "Swapped-Measurements", "postprocess_weight_cat"
    ] = "Include-Fixed-Swap"
    merged_df["bmi"] = merged_df["weight"] / ((merged_df["height"] / 100) ** 2)
    return merged_df


def clean_unit_errors(merged_df):
    """
    This function will look in a DataFrame for rows where the height_cat and weight_cat are set to
    "Unit-Error-High" or "Unit-Error-Low". It will then multiply / divide the height and weight values to convert them.
    It will also create two new columns: postprocess_height_cat and postprocess_weight_cat.
    The values for these columns are copied from the original categories except in the case where
    unit errors are fixed when it is set to "Include-UH" or "Include-UL" respectively.

    Parameters:
    merged_df: (DataFrame) with subjid, height, weight, include_height and include_weight columns

    Returns:
    The cleaned DataFrame
    """
    merged_df["postprocess_height_cat"] = merged_df["height_cat"]
    merged_df["postprocess_height_cat"] = merged_df[
        "postprocess_height_cat"
    ].cat.add_categories(["Include-UH", "Include-UL"])
    merged_df["postprocess_weight_cat"] = merged_df["weight_cat"]
    merged_df["postprocess_weight_cat"] = merged_df[
        "postprocess_weight_cat"
    ].cat.add_categories(["Include-UH", "Include-UL"])
    merged_df.loc[merged_df["height_cat"] == "Unit-Error-Low", "height"] = (
        merged_df.loc[merged_df["height_cat"] == "Unit-Error-Low", "height"] * 2.54
    )
    merged_df.loc[merged_df["height_cat"] == "Unit-Error-High", "height"] = (
        merged_df.loc[merged_df["height_cat"] == "Unit-Error-High", "height"] / 2.54
    )
    merged_df.loc[merged_df["weight_cat"] == "Unit-Error-Low", "weight"] = (
        merged_df.loc[merged_df["weight_cat"] == "Unit-Error-Low", "weight"] * 2.2046
    )
    merged_df.loc[merged_df["weight_cat"] == "Unit-Error-High", "weight"] = (
        merged_df.loc[merged_df["weight_cat"] == "Unit-Error-High", "weight"] / 2.2046
    )
    merged_df.loc[
        merged_df["height_cat"] == "Unit-Error-Low", "postprocess_height_cat"
    ] = "Include-UL"
    merged_df.loc[
        merged_df["height_cat"] == "Unit-Error-High", "postprocess_height_cat"
    ] = "Include-UH"
    merged_df.loc[
        merged_df["weight_cat"] == "Unit-Error-Low", "postprocess_weight_cat"
    ] = "Include-UL"
    merged_df.loc[
        merged_df["weight_cat"] == "Unit-Error-High", "postprocess_weight_cat"
    ] = "Include-UH"
    merged_df["bmi"] = merged_df["weight"] / ((merged_df["height"] / 100) ** 2)
    return merged_df
